nabs/suspenders.py

```diff
@@ -0,0 +1,179 @@
+import asyncio
+import copy
+import logging
+from threading import RLock, Event
+
+from bluesky.plan_stubs import null, wait_for
+from bluesky.preprocessors import plan_mutator
+
+logger = logging.getLogger(__name__)
+
+
+class SuspendPreprocessor:
+    """
+    Base preprocessor that suspends on specific commands based on a signal.
+
+    Parameters
+    ----------
+    signal: ``Signal``
+        The signal to subscribe to, whose value determines when to suspend.
+
+    commands: ``list of str``, optional
+        The commands to suspend on. If omitted, we'll suspend on all commands.
+
+    sleep: ``int`` or ``float``, optional
+        The amount of time to wait after `should_resume` returns ``True``
+        before ending the suspension. If `should_suspend` return ``True`` at
+        any time during this wait period, we will cancel the resumption and
+        return to the suspended state.
+    """
+    def __init__(self, signal, *, commands=None, sleep=0,
+                 pre_plan=null, post_plan=null, follow_plan=null):
+        self._sig = signal
+        self._cmd = commands
+        self._sleep = sleep
```

Comment:
Why are these private?

Reply:
I don't remember. I probably was worried about a malicious user changing them during a run. I can swap them to public if you think adjusting these is useful between runs.

Reply:
I think the `RE` will grab the main thread so not worried about it. I guess they could do it on a `pause`? Not a big enough concern for me. I think they should be public.